### PR TITLE
Update scripts to add and sort rules

### PR DESCRIPTION
**Pull request by JonathanPlasse**
The aim of this PR is to reintroduce the use of `add_rule.py` and `add_plugin.py` in `CONTRIBUTING.md`.
This PR also fixes the sorting of rules, this means that all new code will be sorted correctly.
The CI job to test these scripts has been improved and should test all cases.

### CODE REVIEW

**Comment by JonathanPlasse** [comment]
CI will now fail if the rules are not sorted.
[CI fail](https://github.com/charliermarsh/ruff/actions/runs/5034355069/jobs/9029131682?pr=3747)
[CI success](https://github.com/charliermarsh/ruff/actions/runs/5034442785/jobs/9029257386?pr=3747)

**Comment by JonathanPlasse** [comment]
Only thing remaining, adding the instructions in `CONTRIBUTING.md` to use these scripts.

**Comment by JonathanPlasse** [comment]
Currently there are 24 plugins that uses the old structure:
- [ ] flake8-async
- [ ] flake8-builtins
- [ ] flake8-blind-except
- [ ] flake8-debugger
- [ ] mccabe
- [ ] flake8-tidy-imports
- [ ] flake8-return
- [ ] flake8-gettext
- [ ] flake8-implicit-str-concat
- [ ] flake8-quotes
- [ ] flake8-annotations
- [ ] flake8-future-annotations
- [ ] flake8-2020
- [ ] eradicate
- [ ] flake8-boolean-trap
- [ ] flake8-unused-arguments
- [ ] flake8-datetimez
- [ ] flake8-errmsg
- [ ] flake8-pie
- [ ] flake8-commas
- [ ] flake8-no-pep420
- [ ] flake8-use-pathlib
- [ ] flake8-logging-format
- [ ] flake8-todos

**Comment by JonathanPlasse** [comment]
Ready for review.

**Comment by JonathanPlasse** [comment]
This is ready to merge as is.
Following PR will be needed to:
- Sort all rules
- Enable keeping rules in CI
- Migrate to the new rule architecture

**Comment by JonathanPlasse** [comment]
I finished converting all old architectures to the new one.
I would like both PRs to be merged to avoid having to rebase them.

**Comment by JonathanPlasse** [comment]
- You can now merge this PR or #4589 in any order.

File: scripts/_utils.py [hunk 1]
```diff
@@ -22,3 +35,51 @@ def snake_case(name: str) -> str:
...
 def get_indent(line: str) -> str:
     return re.match(r"^\s*", line).group()  # type: ignore[union-attr]
+
+
+def key_test_case(nb_digit: int) -> Callable[[str], tuple[str, int, str, int]]:
+    def key(line: str) -> tuple[str, int, str, int]:
+        *_, (rule, prefix, code, subcode) = re.findall(
+            (
+                r"(?s)Rule::(.*?),.*?Path::new\("
+                r'(?:"(?:.*?)?([A-Z]+)([0-9]+)?|.+?)(?:.*?)?(_[0-9]+)?(?:.(?:pyi?|txt))?"'
```
Comment (JonathanPlasse) [inline scripts/_utils.py:45]: Would you like some documentation for the regex?

**Comment by charliermarsh** [comment]
Hey @JonathanPlasse - sorry to let this one sit for so long. I really appreciate the work here, but after discussion with some other maintainers, we've opted not to merge this. The main concern is that we're signing up for ongoing maintenance of these Python scripts which are getting increasingly complex and will need to change whenever we change implementation details of the rule files. In other words: they're parsing a lot of Rust code via regex and otherwise, and that Python parsing code will likely need modifications over time. As one example: the current version doesn't handle feature flags, and so is broken in the Ruff rules. I'm sure we could fix that bug, but it's just an example of the kind of thing we'll run into over time. Again, thank you for putting it together and sorry for the extensive delay.

**Pull request closed without merging**